### PR TITLE
Disallow duplicate interval types during parsing

### LINKED ISSUE #3183: `INTERVAL` literal with duplicated interval types should raise error
**Describe the bug**

**To Reproduce**

```bash
❯ select interval '1 second 1 second';
+------------------------------------------------+
| IntervalDayTime("2000")                        |
+------------------------------------------------+
| 0 years 0 mons 0 days 0 hours 0 mins 2.00 secs |
+------------------------------------------------+
1 row in set. Query took 0.000 seconds.
```

note that it's not allowed in postgrseql
```bash
willy=# select interval '1 second 1 second';
ERROR:  invalid input syntax for type interval: "1 second 1 second"
LINE 1: select interval '1 second 1 second';
```

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #3183.

# Rationale for this change

# What changes are included in this PR?

Keep track of interval types found during parsing as bitflag.

Also refactor interval types to use an Enum to better serve above point.

# Are these changes tested?

Added unit test.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/src/parsers.rs [hunk 1]
```diff
@@ -42,56 +82,72 @@ pub fn parse_interval(leading_field: &str, value: &str) -> Result<ScalarValue> {
...
+                value
+            ))
+        })?;
+
+        // Disallow duplicate interval types
+        if used_interval_types & (it as u16) != 0 {
+            return Err(DataFusionError::SQL(ParserError::ParserError(format!(
+                "Invalid input syntax for type interval: {:?}",
+                value
+            ))));
```
Comment (alamb) [inline datafusion/common/src/parsers.rs:121]:
```suggestion
        // Disallow duplicate interval types
        if used_interval_types & (it as u16) != 0 {
            return Err(DataFusionError::SQL(ParserError::ParserError(format!(
                "Invalid input syntax for type interval: {:?}. Repeated type {}",
                value, interval_type
            ))));
```

**Review by alamb — APPROVED** [review]
This looks great to me -- thank you @Jefffrey 

I had a small suggestion to make the error message more specific, and then this PR needs to be rebased, but otherwise I think it is good to go

cc @waynexia  and @waitingkuo

**Review by waynexia — APPROVED** [review]
Looks great! Thanks @Jefffrey and @alamb 👍

**Review by andygrove — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @Jefffrey !

**Pull request merged**